spec/protobuf/message_spec.cr

```diff
@@ -59,4 +59,79 @@ describe Protobuf::Message do
     test4 = TestMessagesProto2::Test4.new(d: [] of Int32)
     test4.to_protobuf.empty?.should be_true
   end
+
+  it "handles oneof in proto2" do
+    # When multiple oneof values are populated
+    # by the ctor only the last survives
```

Comment:
Question about this verbiage. If I have the following protobuf message:

```protobuf
message MyMessage {
  oneof pick_one {
    int32 foo = 1;
    int32 bar = 2;
  }
}
```

And I create these 2 messages:

```crystal
MyMessage.new(foo: 1, bar: 2)
MyMessage.new(bar: 2, foo: 1)
```

Is the expectation that these produce different results?

Reply:
Hmm, yet another good question!

Well, we call the setters at initialization time in FIELDS-order;
`{% for tag, field in FIELDS %}`

So I believe both invocations should produce the same result.
Might be worth adding a test for this although it's not behavior
that should be relied on (strictly speaking we should probably raise
an Exception when someone tries to set multiple values of a oneof
during init).

Reply:
> So I believe both invocations should produce the same result.

They do, it's why I asked 🙂 The hard part is that this would require someone working with message containing a `oneof` to know the order of the declarations in the proto source (assuming we output them in the same order in the Crystal code, which I believe we do). This is unfortunately not very ergonomic.

> we should probably raise an Exception when someone tries to set multiple values of a oneof during init

This might be a good idea. Ideally, we'd use the type system to make this a compile-time error by defining two separate `initialize` methods. Unfortunately, that would only work if all of the members of the `oneof` are different types because Crystal sees `def a(b : Int32)` and `def a(c : Int32)` as the same method.

<img width="532" alt="calling a(b: 42) and a(c: 42) results in the following error on the call with the argument b: missing argument: c." src="https://user-images.githubusercontent.com/108205/126022061-fe422ebe-b141-464e-8072-7281df6a0e08.png">

<details>
  <summary>Click for code</summary>

```crystal
def a(b : Int)
  "B"
end

def a(c : Int)
  "C"
end

a(b: 42)
a(c: 42)
```

</details>

Maybe we could still do that, though. A while back, they added the ability to force some arguments to be passed as named arguments. So in the snippet pasted above, forcing them as named arguments allows the methods to be seen differently:

<img width="712" alt="image" src="https://user-images.githubusercontent.com/108205/126022118-68995b15-4cdf-476a-b309-59c7ea1e60e2.png">

<details>
  <summary>Click for code</summary>

```crystal
def a(*, b : Int)
  "B"
end

def a(*, c : Int)
  "C"
end

a(b: 42)
a(c: 42)
```

</details>

This would require generating all permutations of `initialize` for all `oneof` declarations. So for example:

```proto
message MyMessage {
  oneof one {
    string foo = 1;
    string bar = 2;
  }

  oneof two {
    string baz = 3;
    string quux = 4;
    string omg = 5;
  }
}
```

The resulting Crystal code (after macro expansion) would need to have 6 `initialize` methods:

```crystal
struct MyMessage
  include ::Protobuf::Message

  # ... property declarations ...

  def initialize(*, @foo : String, @baz : String)
  end

  def initialize(*, @foo : String, @quux : String)
  end

  def initialize(*, @foo : String, @omg : String)
  end

  def initialize(*, @bar : String, @baz : String)
  end

  def initialize(*, @bar : String, @quux : String)
  end

  def initialize(*, @bar : String, @omg : String)
  end
end
```

I think this would be an _excellent_ way to enforce `oneof` at compile time, but requires significantly more work.

If we wanted to get fancy, we could even have a pseudo-`initialize` whose only purpose is to show a better error at compile time if you pass both. But that's probably not unnecessary. The compiler error is probably enough for someone using the message to look at the definition and realize "oh, right, I can only specify one".

Reply:
Holy ravioli, that looks complicated! 🤯 😅

What exactly is this trying to solve, though? 🤔 

Just to be clear: I think incoming messages are not supposed to have multiple values for a oneof.
(the current version of protobuf.cr will indeed create such messages but I don't think they are legal)

From what I gathered from the docs I thought it should be enough to enforce their
exclusivity at object creation time (i.e. in the constructor or when a setter is called).

But could be I missed something 🤔

Reply:
> Holy ravioli, that looks complicated! 🤯 😅

It does when the macros are expanded, but it would all be abstracted from anyone reading the generated code.

The implementation could be relatively simple:

- Are there any `oneof`s in this message?
  - If yes: iterate over them in a nested loop generating a new `initialize` method for each one
  - If no: generate the one `initialize`

The downside is that it's a breaking change for anyone using positional arguments to `initialize`, which I believe we do in the test suite.

> What exactly is this trying to solve, though? 🤔

Setting two values in a `oneof` group.  You have to know you can only supply one key from each `oneof` group no matter what the implementation is, but right now we're just not checking your work on that. Your implementation would check your work, which is awesome, but if you get it wrong we just overwrite all but one of your values. You may have no idea that's happened.

Your idea of raising an exception _will_ let them know, which is arguably a huge improvement over silently dropping values because it lets them know that their code has a bug, but it happens at run time. People will likely discover the bug in production. Generating multiple `initialize` methods will move that exception to compile time. They won't even be able to deploy the bug because it won't be a valid Crystal program.

> Just to be clear: I think incoming messages are not supposed to have multiple values for a oneof.
> (the current version of protobuf.cr will indeed create such messages but I don't think they are legal)

Agreed.
 
> From what I gathered from the docs I thought it should be enough to enforce their
> exclusivity at object creation time (i.e. in the constructor or when a setter is called).
> 
> But could be I missed something 🤔

We don't have a guarantee when we're deserializing anything coming in from an untrusted host, but the risk there is probably relatively low. It should be pretty rare that nobody using of a particular message is using a library that checks that unless they're all using the same Protobuf library to serialize/deserialize — at which point, Protobuf is overkill and they should probably be using something else.

Reply:
> breaking change for anyone using positional arguments

Hmm yes, as mentioned in my prev. comment (sorry I had edited it later on - it got late that night)
I'm skeptical about the benefits we'd gain from it.

Having an explicit `Foo.new(a: 1, b:2)` in the code (where only one value should
be provided) seems like a rare case and should usually be easily caught in testing.

I think the other protobuf impls (at least the ones I've looked at)
don't go to that length either.

But in any case, your scrutiny has led me to an actual bug in the current impl! 🐞 

Now fixed [here](https://github.com/jeromegn/protobuf.cr/pull/40/commits/4e8c566541c58aeaa2ec424a15a5bcffd186816a).
Tldr: We **do** need to call the setters after deserialisation or the instance var
that identifies which oneof has a value doesn't get set.

I've also added a test for it and deduplicated them [here](https://github.com/jeromegn/protobuf.cr/pull/40/commits/0c22b974d75de8446055ae34c92b40b032083853).

With this fix I believe the current impl now works as intended and now it
also applies the "keep only first value"-behavior to incoming messages.

I'm not opposed to refactoring into multiple ctor's as you propose
but am still not quite convinced it's worth it. I feel like it may also bring
some undesirable side-effects such as potentially confusing compiler errors
as messages with many fields may end up with dozens of ctors. 🤔